File: solana_launchpads/go/launchpad/raydium_launchlab.go
package launchpad

// Raydium LaunchLab -- program LanMV9sAd7wArD4vJFi2qDdfnVhFxYSUg6eADduJ3uj
//
// Multi-tenant, and unlike pump.fun almost everything is on the pool itself:
// supply, the sellable amount, the raise target, and the reserve pair. Only two
// things come from elsewhere -- which curve maths applies (GlobalConfig) and
// the platform's fee split (PlatformConfig).
//
// The three curve types differ in what virtual_base and virtual_quote mean:
//
//	constant product : virtual liquidity; price = (vQuote+realQuote)/(vBase-realBase)
//	fixed price      : price = virtual_quote / virtual_base, flat for the whole curve
//	linear           : virtual_base is the slope in Q64; price = a*sold/2^64
//
// Note the virtual reserves are written once at pool creation and never move --
// only the real ones accumulate. That is what makes the k check below work, and
// it also means a virtual reserve that differs between two pools sharing one
// config is a pipeline artefact rather than chain state.

const (
	// CurveTypeConstantProduct and friends are GlobalConfig.CurveType.
	CurveTypeConstantProduct uint8 = 0
	CurveTypeFixedPrice      uint8 = 1
	CurveTypeLinear          uint8 = 2

	// LaunchLab fee rates are in hundredths of a basis point (1e-6).
	launchLabFeeDenominator = 1_000_000

	// PoolStatus values.
	poolStatusFunding  uint8 = 0
	poolStatusMigrate  uint8 = 1
	poolStatusMigrated uint8 = 2
)

// RaydiumLaunchlabParams carries the per-program and per-platform inputs that
// a pool alone does not supply.
type RaydiumLaunchlabParams struct {
	CurveType uint8
	// TradeFeeRate, PlatformFeeRate and CreatorFeeRate are in 1e-6 units.
	TradeFeeRate    uint64
	PlatformFeeRate uint64
	CreatorFeeRate  uint64
	// MigrateFee is the config's default, used only when the pool's own
	// migrate_fee is zero.
	MigrateFee uint64
	// FromChain is false when the config was not read from a config account,
	// in which case CurveType is an assumption rather than a fact.
	FromChain bool
}

// NewRaydiumLaunchlabParams reads what is needed from the GlobalConfig. Pass
// the PlatformConfig separately via WithPlatform when it is available; without
// it the fee is understated but every price is still correct.
func NewRaydiumLaunchlabParams(g *RaydiumLaunchlabGlobalConfig) RaydiumLaunchlabParams {
	return RaydiumLaunchlabParams{
		CurveType:    g.CurveType,
		TradeFeeRate: g.TradeFeeRate,
		MigrateFee:   g.MigrateFee,
		FromChain:    true,
	}
}

// WithPlatform folds a platform's fee split into the params.
func (p RaydiumLaunchlabParams) WithPlatform(pc *RaydiumLaunchlabPlatformConfig) RaydiumLaunchlabParams {
	p.PlatformFeeRate = pc.FeeRate
	p.CreatorFeeRate = pc.CreatorFeeRate
	return p
}

// CurveTypeName renders the curve type for reporting.
func (p RaydiumLaunchlabParams) CurveTypeName() string {
	switch p.CurveType {
	case CurveTypeConstantProduct:
		return "constant_product"
	case CurveTypeFixedPrice:
		return "fixed_price"
	case CurveTypeLinear:
		return "linear"
	}
	return "unknown"
}

// Family maps the curve type onto the shared vocabulary.
func (p RaydiumLaunchlabParams) Family() CurveFamily {
	switch p.CurveType {
	case CurveTypeConstantProduct:
		return FamilyConstantProductVirtual
	case CurveTypeFixedPrice:
		return FamilyFixedPrice
	case CurveTypeLinear:
		return FamilyLinear
	}
	return FamilyUnknown
}

// Metrics turns one decoded pool into the unified answer.
func (p RaydiumLaunchlabParams) Metrics(pool *RaydiumLaunchlabPoolState, truncatedAt string) Metrics {
	baseDec := int(pool.BaseDecimals)
	quoteDec := int(pool.QuoteDecimals)

	paramsSource := SourceOnchainConfig
	if !p.FromChain {
		paramsSource = SourceFallback
	}

	m := Metrics{
		Launchpad:     "raydium_launchlab",
		ProgramID:     RaydiumLaunchlabProgramID,
		Family:        p.Family(),
		CurveType:     p.CurveTypeName(),
		BaseMint:      pool.BaseMint,
		QuoteMint:     pool.QuoteMint,
		BaseDecimals:  baseDec,
		QuoteDecimals: quoteDec,
		TotalSupply:   UIAmount(pool.Supply, baseDec),
		TokensForSale: UIAmount(pool.TotalBaseSell, baseDec),
		TokensSold:    UIAmount(pool.RealBase, baseDec),
		RaiseTarget:   UIAmount(pool.TotalQuoteFundRaising, quoteDec),
		Raised:        UIAmount(pool.RealQuote, quoteDec),
		Complete:      pool.Status >= poolStatusMigrate,
		Migrated:      pool.Status == poolStatusMigrated,
		PriceSource:   SourceOnchainState,
		ParamsSource:  paramsSource,
		TruncatedAt:   truncatedAt,
	}

	switch p.CurveType {
	case CurveTypeConstantProduct:
		m.LaunchPrice = PriceUI(CPPriceRaw(pool.VirtualQuote, pool.VirtualBase), baseDec, quoteDec)
		if pool.VirtualBase > pool.RealBase {
			m.CurrentPrice = PriceUI(
				CPPriceRaw(pool.VirtualQuote+pool.RealQuote, pool.VirtualBase-pool.RealBase),
				baseDec, quoteDec)
		}
		m.Violations = CheckLaunchLabConstantProduct(
			pool.VirtualBase, pool.VirtualQuote, pool.RealBase, pool.RealQuote)
	case CurveTypeFixedPrice:
		m.LaunchPrice = PriceUI(CPPriceRaw(pool.VirtualQuote, pool.VirtualBase), baseDec, quoteDec)
		m.CurrentPrice = m.LaunchPrice
	case CurveTypeLinear:
		// A linear curve opens at price zero and rises with tokens sold.
		m.LaunchPrice = 0
		m.CurrentPrice = PriceUI(
			LaunchLabLinearPriceRaw(pool.VirtualBase, pool.RealBase), baseDec, quoteDec)
	}

	// Graduation price is what the migrated AMM pool opens at: the raise, less
	// the migration fee, spread over the tokens held back for migration.
	migrateFee := pool.MigrateFee
	if migrateFee == 0 {
		migrateFee = p.MigrateFee
	}
	migrateTokens := int64(pool.Supply) - int64(pool.TotalBaseSell) -
		int64(pool.VestingSchedule.TotalLockedAmount)
	if migrateTokens > 0 && pool.TotalQuoteFundRaising > migrateFee {
		raw := float64(pool.TotalQuoteFundRaising-migrateFee) / float64(migrateTokens)
		m.GraduationPrice = PriceUI(raw, baseDec, quoteDec)
	}

	rates := p.TradeFeeRate + p.PlatformFeeRate + p.CreatorFeeRate
	m.FeeBps = float64(rates) / launchLabFeeDenominator * 10_000

	if len(m.Violations) > 0 {
		m.PriceSource = "SUSPECT"
	}
	m.fillDerived()
	return m
}

// CheckLaunchLabConstantProduct tests a LaunchLab pool's reserves against the k
// its virtual pair fixes.
//
// LaunchLab writes virtual_base and virtual_quote once at pool creation and
// never touches them again; only the real reserves accumulate. So the effective
// product
//
//	(virtual_base - real_base) * (virtual_quote + real_quote)
//
// must stay at virtual_base * virtual_quote, up to the same flooring dust that
// only ever moves it up. A pool that misses it has real and virtual reserves
// that did not come from the same read.
func CheckLaunchLabConstantProduct(virtualBase, virtualQuote, realBase, realQuote uint64) []Violation {
	if virtualBase == 0 || virtualQuote == 0 {
		return nil
	}
	if realBase >= virtualBase {
		return []Violation{{
			Name: "real_base_exceeds_virtual",
			Detail: "real_base is not below virtual_base, which would empty the curve; " +
				"the two did not come from one read",
			Severity: SeverityError,
		}}
	}
	return CheckConstantProductPair(
		virtualBase-realBase, virtualQuote+realQuote, virtualBase, virtualQuote)
}

File: solana_launchpads/launchpad_decoder/adapters/raydium_launchlab.py
"""Raydium LaunchLab pools (LetsBonk.fun, Cook.meme, Raydium's own UI, ...).

Program: LanMV9sAd7wArD4vJFi2qDdfnVhFxYSUg6eADduJ3uj

LaunchLab is multi-tenant, and everything needed is on the pool itself:
`supply`, `total_base_sell`, `total_quote_fund_raising`, the virtual/real
reserve pair, plus pointers to the `GlobalConfig` (which curve type this pool
uses) and the `PlatformConfig` (which platform launched it).

The three curve types are implemented exactly as `raydium-sdk-V2` does:

* constant product -- price = (virtualB + realB) / (virtualA - realA)
* fixed price      -- price = virtualB / virtualA, flat for the whole curve
* linear           -- price = virtualA * realA / 2^64  (virtualA is the slope)
"""

from __future__ import annotations

from typing import Any, Dict, Optional, Tuple

from .. import curves
from ..types import CurveFamily, LaunchMetrics, ValueSource, price_raw_to_ui, ui_amount
from .base import Adapter, DecodeContext

CURVE_TYPES = {0: "constant_product", 1: "fixed_price", 2: "linear"}
FAMILY_BY_CURVE = {
    0: CurveFamily.CONSTANT_PRODUCT_VIRTUAL,
    1: CurveFamily.FIXED_PRICE,
    2: CurveFamily.LINEAR,
}
POOL_STATUS = {0: "funding", 1: "awaiting_migration", 2: "migrated"}

#: `trade_fee_rate` and friends are denominated in hundredths of a bip (1e-6)
FEE_RATE_DENOMINATOR = 1_000_000


class RaydiumLaunchLabAdapter(Adapter):
    key = "raydium_launchlab"

    def _configs(self, ctx: DecodeContext, state: Dict[str, Any]) -> Tuple[dict, dict, int]:
        global_cfg: dict = {}
        platform_cfg: dict = {}
        slot = 0
        for field, target in (("global_config", "g"), ("platform_config", "p")):
            address = state.get(field)
            if not address:
                continue
            name, decoded, cfg_slot = ctx.configs.fetch(
                ctx.spec.program_id, address, ctx.schema
            )
            if decoded is None:
                continue
            slot = max(slot, cfg_slot)
            if target == "g":
                global_cfg = decoded
            else:
                platform_cfg = decoded
        return global_cfg, platform_cfg, slot

    def decode(
        self,
        ctx: DecodeContext,
        account_name: str,
        state: Dict[str, Any],
        address: Optional[str] = None,
    ) -> LaunchMetrics:
        m = self.new_metrics(ctx, account_name, address, state)
        global_cfg, platform_cfg, cfg_slot = self._configs(ctx, state)

        m.base_mint = state.get("base_mint")
        m.quote_mint = state.get("quote_mint") or ctx.spec.default_quote_mint
        m.creator = state.get("creator")
        m.base_decimals = state.get("base_decimals")
        m.quote_decimals = state.get("quote_decimals")
        if m.base_decimals is None:
            m.base_decimals = ctx.decimals_for(m.base_mint)
        if m.quote_decimals is None:
            m.quote_decimals = ctx.decimals_for(m.quote_mint)

        curve_type = global_cfg.get("curve_type")
        if curve_type is None:
            curve_type = 0
            m.warn("GlobalConfig unavailable; assuming constant-product curve_type=0")
        m.curve_type = CURVE_TYPES.get(curve_type, f"unknown({curve_type})")
        m.curve_family = FAMILY_BY_CURVE.get(curve_type, CurveFamily.UNKNOWN)

        supply = state.get("supply")
        total_base_sell = state.get("total_base_sell")
        total_raise = state.get("total_quote_fund_raising")
        virtual_a = state.get("virtual_base")
        virtual_b = state.get("virtual_quote")
        real_a = state.get("real_base") or 0
        real_b = state.get("real_quote") or 0
        migrate_fee = state.get("migrate_fee") or global_cfg.get("migrate_fee") or 0
        vesting = state.get("vesting_schedule") or {}
        locked = vesting.get("total_locked_amount") or 0

        m.total_supply = self.param(
            ui_amount(supply, m.base_decimals), ValueSource.ONCHAIN_STATE, ctx.slot
        )
        m.tokens_for_sale = self.param(
            ui_amount(total_base_sell, m.base_decimals), ValueSource.ONCHAIN_STATE, ctx.slot
        )
        m.tokens_sold = self.param(
            ui_amount(real_a, m.base_decimals), ValueSource.ONCHAIN_STATE, ctx.slot
        )
        m.raise_target_quote = self.param(
            ui_amount(total_raise, m.quote_decimals),
            ValueSource.ONCHAIN_STATE,
            ctx.slot,
            "total_quote_fund_raising",
        )
        m.raised_quote = self.param(
            ui_amount(real_b, m.quote_decimals), ValueSource.ONCHAIN_STATE, ctx.slot
        )

        # -- prices, per curve type ----------------------------------------
        launch_raw = current_raw = None
        if curve_type == 0 and virtual_a and virtual_b:
            launch_raw = curves.cp_price_raw(virtual_b, virtual_a)
            denominator = virtual_a - real_a
            current_raw = curves.cp_price_raw(virtual_b + real_b, denominator)
        elif curve_type == 1 and virtual_a and virtual_b:
            launch_raw = curves.fixed_price_raw(virtual_b, virtual_a)
            current_raw = launch_raw
        elif curve_type == 2 and virtual_a:
            launch_raw = 0.0  # a linear curve starts at price zero
            current_raw = curves.launchlab_linear_price_raw(virtual_a, real_a)

        m.launch_price_quote = self.param(
            price_raw_to_ui(launch_raw, m.base_decimals, m.quote_decimals),
            ValueSource.ONCHAIN_STATE,
            ctx.slot,
            f"{m.curve_type} initial price",
        )
        m.current_price_quote = self.param(
            price_raw_to_ui(current_raw, m.base_decimals, m.quote_decimals),
            ValueSource.ONCHAIN_STATE,
            ctx.slot,
        )

        # Graduation price is the price the migrated AMM pool opens at:
        # (raise - migrate fee) spread over the tokens reserved for migration.
        if supply is not None and total_base_sell is not None and total_raise is not None:
            migrate_tokens = supply - total_base_sell - locked
            if migrate_tokens > 0:
                end_raw = (total_raise - migrate_fee) / migrate_tokens
                m.graduation_price_quote = self.param(
                    price_raw_to_ui(end_raw, m.base_decimals, m.quote_decimals),
                    ValueSource.DERIVED,
                    ctx.slot,
                    "(total_quote_fund_raising - migrate_fee) / (supply - total_base_sell - locked)",
                )

        status = state.get("status")
        m.complete = self.param(
            None if status is None else status >= 1, ValueSource.ONCHAIN_STATE, ctx.slot,
            POOL_STATUS.get(status, ""),
        )
        m.migrated = self.param(
            None if status is None else status == 2, ValueSource.ONCHAIN_STATE, ctx.slot
        )

        trade_fee = global_cfg.get("trade_fee_rate")
        platform_fee = platform_cfg.get("fee_rate")
        creator_fee = platform_cfg.get("creator_fee_rate")
        rates = [r for r in (trade_fee, platform_fee, creator_fee) if r is not None]
        if rates:
            m.fee_bps = self.param(
                sum(rates) / FEE_RATE_DENOMINATOR * 10_000,
                ValueSource.ONCHAIN_CONFIG,
                cfg_slot,
                "trade + platform + creator fee rates (1e-6 units) as bps",
            )

        platform_name = _platform_name(platform_cfg)
        if platform_name:
            m.raw_state = {**state, "_platform_name": platform_name}

        return self.finish(m)


def _platform_name(platform_cfg: dict) -> str:
    raw = platform_cfg.get("name")
    if isinstance(raw, (list, tuple)):
        try:
            return bytes(int(b) for b in raw).rstrip(b"\x00").decode("utf-8", "replace")
        except (TypeError, ValueError):
            return ""
    return raw if isinstance(raw, str) else ""
